Gravatar Android Demo App Internal Distribution

wzieba opened the pull request:
### Description

This PR configures a CI step for building the demo app, share it to S3 bucket and comment on a PR with link/qr code to download.

wzieba committed:
Add build_and_upload_prototype_build lane

wzieba committed:
Add Buildkite pipeline step to build and upload prototype

iangmaia commented on .buildkite/pipeline.yml:
```diff
@@ -82,3 +82,10 @@ steps:
 
+  - label: "Prototype Builds"
+    if: "build.pull_request.id != null"
+    command: |
+      ".buildkite/commands/prototype-build.sh"
```
```suggestion
    command: .buildkite/commands/prototype-build.sh
```

iangmaia committed:
Update .rubocop.yml config

dangermattic commented on fastlane/Fastfile:
```diff
@@ -54,5 +55,63 @@ platform :android do
+
+    desc 'Builds a prototype build and uploads it to S3'
+    lane :build_and_upload_prototype_build do
+      unless ENV['BUILDKITE_ARTIFACTS_S3_BUCKET']
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/Gravatar-SDK-Android/blob/a5e3bf10b0a8bbe71470202d5a27bdcf3c1ae6a2/fastlane/Fastfile#L61"></span>Style/IfUnlessModifier: Favor modifier <code>unless</code> usage when having a single-line body. Another good alternative is the usage of control flow <code>&amp;&amp;</code>/<code>||</code>.</td>
    </tr>
  </tbody>
</table>

iangmaia committed:
Rubocop update & correct violations

iangmaia committed:
Update .buildkite/pipeline.yml

iangmaia committed:
Uodate to latest a8c-ci-toolkit

iangmaia committed:
Update Bundler configuration

iangmaia committed:
Adjust Fastfile lane `build_and_upload_prototype_build`

dangermattic commented on fastlane/Fastfile:
```diff
@@ -55,4 +56,60 @@ platform :android do
+
+      "#{branch}-#{commit}"
+    end
+  end  
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/Gravatar-SDK-Android/blob/280d08f3a066e0b5a43be9026426561e2b4debbb/fastlane/Fastfile#L114"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -55,4 +56,60 @@ platform :android do
+      bucket: 'a8c-apps-public-artifacts',
+      key: "gravatar-demo-prototype-build-#{generate_prototype_build_number}.apk",
+      file: lane_context[SharedValues::GRADLE_ALL_APK_OUTPUT_PATHS].first,
+      skip_if_exists: true
```
Side note: `skip_if_exists` parameter is deprecated, use `if_exists: skip` instead.

AliSoftware replied:
@wzieba Sorry for the typo here, that was supposed to be `if_exists: :skip` with a `:` in front of the `skip` keyword (to make it a Ruby symbol, instead of making it look like the name of a variable being referenced)

wzieba committed:
Extract "Publish to S3" step to a separate command

In next step, I'll include ENV export.

Citing @AliSoftware

> Though, for security reasons, you’d want to first move the command called by that Publish to S3 in a dedicated .buildkite/commands/publish-to-s3.sh script.
  Not only so that the value of those secret env vars is not exposed in the logs (e.g. Buildkite exposes the content of the BUILDKITE_COMMAND in the env vars so that would expose the values if you don’t add that indirection)
  But also so that the ${A8C_LIBS_AWS_ACCESS_KEY} values are not resolved too soon (at pipeline upload time) but instead only when the command is run

wzieba committed:
Export AWS credentials in `publish_to_s3` command (see comment for the context)

wzieba committed:
Address Quote this to prevent word splitting (SC2046).

wzieba committed:
`skip_if_exists` parameter is deprecated

wzieba commented:
Does it look good to you @AliSoftware @iangmaia ? If yes, I'll update our secrets and re-run build here

AliSoftware commented on .buildkite/commands/publish_to_s3.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash -eu
+
+# The A8C_LIBS_* credentials are for the `a8c-libs` S3 bucket in a separate `a8c` account.
+# We use distinct names to prevent conflicts with the AWS credentials of our `a8c-apps` account (used for all our other AppsInfra buckets).
```
Re-reading this in context, here's another suggestion you might find even clearer for that comment here?

```suggestion
# The A8C_LIBS_* credentials are for the `a8c-libs` S3 bucket in which we want to publish the SDK artifact, and which lives in a separate `a8c` AWS account.
# We don't use the standard AWS env names here to prevent conflicts with the AWS credentials of our main `a8c-apps` account (used for all our other AppsInfra S3 buckets).
```

wdyt?

wzieba replied:
Yes, it's very clear now 👍 thanks!

AliSoftware commented on .buildkite/commands/publish_to_s3.sh:
```diff
@@ -0,0 +1,17 @@
+# The A8C_LIBS_* credentials are for the `a8c-libs` S3 bucket in a separate `a8c` account.
+# We use distinct names to prevent conflicts with the AWS credentials of our `a8c-apps` account (used for all our other AppsInfra buckets).
+
+export AWS_ACCESS_KEY=${A8C_LIBS_AWS_ACCESS_KEY}
+export AWS_SECRET_KEY=${A8C_LIBS_AWS_SECRET_KEY}
```
```suggestion
export AWS_ACCESS_KEY="${A8C_LIBS_AWS_ACCESS_KEY}"
export AWS_SECRET_KEY="${A8C_LIBS_AWS_SECRET_KEY}"
```

AliSoftware commented on .buildkite/pipeline.yml:
```diff
@@ -61,16 +61,7 @@ steps:
     if: build.branch == "trunk" || build.tag != null
 
   - label: "Publish to S3"
```
Nitpick: I wonder if it'd help clarity (i.e. avoid confusing this with the publication of prototype builds) to rename this step to:
```suggestion
  - label: "Publish SDK to S3"
```

AliSoftware reviewed (approved):
@wzieba  Yep this approach LGTM 👍

wzieba committed:
Improve code comment and code quality suggestions

wzieba committed:
Fix `if_exists` typo

wzieba committed:
Fix command name typo

wzieba committed:
Remove parenthesis from `prepare_to_publish_to_s3_params`

Possible fix for "Either tagName or both branch name and sha1 command line arguments needs to be provided!"

wzieba committed:
Fix invalid name of project in `comment_on_pr` Fastlane action

wzieba committed:
Merge branch 'trunk' into demo_app_in_prs

wzieba committed:
Improve "build in progress" PR comment message
